File: src/main/java/me/creepinson/creepinoutils/util/BlockTileData.java
package me.creepinson.creepinoutils.util;

import net.minecraft.nbt.CompoundNBT;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;

public class BlockTileData extends BlockData {

    private TileEntity tileEntity;

    public TileEntity getTileEntity() {
        return tileEntity;
    }

    public void setTileEntity(TileEntity tileEntity) {
        this.tileEntity = tileEntity;
    }

    public BlockTileData(World world) {
        super(world);
    }

    @Override
    public CompoundNBT serializeNBT() {
        CompoundNBT nbt = super.serializeNBT();
        if (tileEntity != null && !tileEntity.isRemoved()) {
            nbt.setTag("tile", tileEntity.serializeNBT());
        }
        return nbt;
    }

    @Override
    public void deserializeNBT(CompoundNBT nbt) {
        super.deserializeNBT(nbt);
        if (nbt.hasKey("tile")) {
            this.tileEntity = TileEntity.create(getWorld(), nbt.getCompoundTag("tile"));
        }
    }
}

File: src/main/kotlin/com/theoparis/creepinoutils/util/BlockTileData.kt
package com.theoparis.creepinoutils.util

import net.minecraft.nbt.CompoundNBT
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World

class BlockTileData(world: World?) : BlockData(world) {
    var tileEntity: TileEntity? = null
    override fun serializeNBT(): CompoundNBT {
        val nbt = super.serializeNBT()
        if (tileEntity != null && !tileEntity!!.isRemoved)
            nbt.put("tile", tileEntity!!.serializeNBT())

        return nbt
    }

    override fun deserializeNBT(nbt: CompoundNBT) {
        super.deserializeNBT(nbt)
        if (nbt.contains("tile"))
            tileEntity = TileEntity.readTileEntity(state, nbt.getCompound("tile"))
    }
}